core/src/com/unciv/ui/utils/KeyPressDispatcher.kt

```diff
@@ -61,14 +82,97 @@ class KeyPressDispatcher: HashMap<KeyCharAndCode, (() -> Unit)>() {
     operator fun contains(code: Int) = this.contains(KeyCharAndCode(code))
     fun remove(code: Int) = this.remove(KeyCharAndCode(code))
 
+    // access by KeyCharAndCode
+    operator fun set(key: KeyCharAndCode, action: () -> Unit) {
+        super.put(key, action)
+        checkInstall()
+    }
+    override fun remove(key: KeyCharAndCode): (() -> Unit)? {
+        val result = super.remove(key)
+        checkInstall()
+        return result
+    }
+
+    override fun toString(): String {
+        return (if (name==null) "" else "$name.") +
+            "KeyPressDispatcher(" + keys.joinToString(limit = 6){ it.toString() } + ")"
+    }
+
+    /** Removes all of the mappings, including a checkpoint if set. */
     override fun clear() {
         checkpoint = setOf()
         super.clear()
+        checkInstall()
     }
+
+    /** Set a checkpoint: The current set of keys will not be removed on a subsequent [revertToCheckPoint] */
     fun setCheckpoint() {
         checkpoint = keys.toSet()
     }
+    /** Revert to a checkpoint: Remove all mappings except those that existed on a previous [setCheckpoint] call.
+     * If no checkpoint has been set, this is equivalent to [clear] */
     fun revertToCheckPoint() {
         keys.minus(checkpoint).forEach { remove(it) }
+        checkInstall()
+    }
+
+    /** install our [EventListener] on a stage with optional inhibitor
+     * @param   stage The [Stage] to add the listener to
+     * @param   hasOpenPopups An optional lambda - when it returns true all keys are ignored
+     */
+    fun install(stage: Stage, name: String? = null, hasOpenPopups: (() -> Boolean)? = null) {
+        this.name = name
+        if (installStage != null) uninstall()
+        listener =
+            object : InputListener() {
+                override fun keyTyped(event: InputEvent?, character: Char): Boolean {
+                    val key = KeyCharAndCode(event, character)
+
+                    if (!contains(key) || (hasOpenPopups?.invoke() == true))
+                        return super.keyTyped(event, character)
```

Comment:
This is just 'return false', but this makes it look like the super actually has extra logic - please change to reduce apparent complexity

Reply:
Uh, that's not how I see it - I mean the line is even auto-inserted by the IDE like that. Point is, super is part of an API and out of our control and subject to change - potentially. remotely possible. perhaps. So we treat it like documented - if we feel responsible we handle the event if not we let it bubble. When I think of dotnet UI frameworks like Windows Forms or WPF it was always quite vital to handle only what's needed and pass on the rest... Am I glad I've avoided those for a while now.

... or am I missing something?

Reply:
I understand what you're saying. I don't necessarily agree, but I'm willing to accept this is a difference of opinion that's not critical :)